File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.cs
public class Solution {
    // A brute force solutiuon would be to start a substring at every index using a nested loop.
    // We can check if the current substring contains a duplicate by storing its chars in a HashSet
    // for O(1) lookups. If it does, store its length and move to the next index.
    // Time Complexity would be O(n * m) where n is the number of chars in the string and m is the number
    // of unique characters stored in the HashSet
    // Space Complexity is O(m) where m is the number of unique characters
    public int LengthOfLongestSubstring(string s) {
        var length = 0;

        for (var i = 0; i < s.Length; i++) {
            var hashSet = new HashSet<char>();

            for (var j = i; j < s.Length; j++) {
                if(hashSet.Contains(s[j])) break;
                hashSet.Add(s[j]);
                length = Math.Max(hashSet.Count, length);
            }
        }

        return length;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-7.js
class Solution {
    // I will use a sliding window to get substring and then 
    // use a hashmap to get the frequency count of that 
    // substring. I will do so because we need to figure 
    // how many character we can replace in that substring
    // to have all distinct characters. 
    // And to do so we need to get the least freq character
    // in the curr substring which would be the size of the window
    // minus the most freq character. As long as char to replace
    // is less than or equal to k then we have valid window
    // which we can expand. Otherwise shift the window to the right.

    // Time Complexity would O(n)
    // Space Complexity would be O(m)
    // Where n is the length of the string and m is the number
    // of unique characters. 

    // Brute force would be to compare every char with every
    // other char in the string using a nested loop
    // That would be O(n^2) time complexity and O(1) space
    // complexity
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        const hashMap = new Map();
        let res = 0;
        let l = 0;
        
        for (let r = 0; r < s.length; r++) {
            hashMap.set(s[r], (hashMap.get(s[r]) || 0) + 1);

            // invalid window
            while (r - l + 1 - Math.max(...hashMap.values()) > k) {
                hashMap.set(s[l], hashMap.get(s[l]) - 1);
                l++;
            }

            res = Math.max(res, r - l + 1);
        }

        return res;
    }
}
